2016/proposals/result-aggregation-server-garg-prerit.md

```diff
@@ -0,0 +1,200 @@
+* User email address ( optional ) 
+ 
+###Server and API 
+An API is to be developed to store the results generated by the test script. I shall be using Python's web framework [FLASK](http://flask.pocoo.org/) for this purpose. Reason being, flask is a [microframework](https://en.wikipedia.org/wiki/Microframework) with minimal complexity, and can be easily relied upon, as far as less complex projects are concerned. As we don't require a lot of endpoints or functionalities in this API, flask is the most suitable option as it seems. This would involve : 
+* Modification of the installation-test-scrips and giving the user an option to send the data to the server by making a [POST](https://en.wikipedia.org/wiki/POST_(HTTP)) request to the API endpoint with payload of all the information collected. Code Snippet : 
+``` 
+import requests 
+ 
+headers = { 
+    "Content-Type": "application/json", 
+} 
+HOST = "http://127.0.0.1:5000" 
+os_end_point = "/system_info/" 
+data = {"system_dist" : system_dist , "version": version} 
+uri = "%s%s" % (HOST, os_end_point) 
+response = requests.post(uri , data = json.dumps(data) , headers = headers) 
+``` 
+* This API will follow the [REST](https://en.wikipedia.org/wiki/Representational_state_transfer) standard, and all the data shall be send in [json](http://www.json.org/) to the API. 
+ 
+* Setting up the **Database**. As mentioned in the description of the idea, [SQLite](https://www.sqlite.org/) is preferred, but it suffers some problems in case 40-50 people send the data simultaneously. After going through the comparison given [here](https://www.digitalocean.com/community/tutorials/sqlite-vs-mysql-vs-postgresql-a-comparison-of-relational-database-management-systems), I would personally prefer [MySQL](https://www.mysql.com/) or [PostgreSQL](http://www.postgresql.org/) in this project. But that can be decided later as per the discussion with mentors. 
+ 
+* Modelling the information collected at the client end. There shall be Tables in the database for successful installs, failed installs, and diagnose information. All shall be reference with a **unique id**, which will be used as **foreign key** in all the tables, to identify a particular user associated with that entry in the DB. Models shall be defined as follows: 
+``` 
+from app import db 
+class Successful(db.Model): 
+  id = db.Column(db.Integer, primary_key=True) 
+  package_name = db.Column(db.String(128)) 
+  version = db.Column(db.String(128)) 
+``` 
+###Database Schema
+* User shall be prompted to enter the **workshop_id** and **email_id** before sending the data to the server, as it would enable the server to associate each result with the workshop, and summarize conclusions workshop specific and also overall.
+* Concatenation of **workshop_id** and **email_id** as **workshopid_email_id** would give us a unique id per user per workshop, and thus would prevent duplicate entries in the table for the same student attending the same workshop.
```

Comment:
@pbanaszkiewicz yes i do have an id as primary key. But I added workshopid_email_id to prevent duplicate entries for one student attending a workshop. They might run the installation script more than once, and then there shall be duplicate entries for a particular student. To prevent that we can have this unique id, so next time when the script runs, data shall be updated, instead of inserting.